[mlir][debug] Inherit DISubprogramAttr from DILocalScopeAttr.

abidh opened the pull request:
As mentioned in https://github.com/llvm/llvm-project/pull/154926, `DISubprogramAttr` is inherited from `DIScopeAttr` while in llvm, the `DISubprogram` inherits from `DILocalScope`. This change corrects the hierarchy.

abidh committed:
[mlir][debug] Inherit DISubprogramAttr from DILocalScopeAttr.

This better matches the hierarchy in llvm.

gysit reviewed (approved):
Thanks for the followup.

LGTM

Can you update the base class for DILexicalBlockAttr and DILexicalBlockFileAttr as well? It seems like this is an inconsistency from a previous refactor.

abidh committed:
Handle review comments.

The DILexicalBlockAttr and DILexicalBlockFileAttr also inherits from
the DILocalScopeAttr.

abidh commented on mlir/include/mlir/Dialect/LLVMIR/LLVMAttrDefs.td:
```diff
@@ -638,7 +638,7 @@ def LLVM_DILocalVariableAttr : LLVM_Attr<"DILocalVariable", "di_local_variable",
 def LLVM_DISubprogramAttr : LLVM_Attr<"DISubprogram", "di_subprogram",
                                       [LLVM_DIRecursiveTypeAttrInterface],
-                                      "DIScopeAttr"> {
+                                      "DILocalScopeAttr"> {
```
Looking at the code a bit more, the `DILexicalBlockAttr` and `DILexicalBlockFileAttr` have `scope` parameters which are `DIScopeAttr` but probably needs to be  `DILocalScopeAttr`. I tried a few quick changes but they soon became more invasive. I would leave them for later PR.

gysit replied:
Ok 👍
